chore: implement `throwIfDisposed`

jrandolf-google opened the pull request:
If an object is already disposed, the `throwIfDisposed` decorator ensures that methods decorated with it will throw when the parent object is disposed.

jrandolf committed:
chore: implement `throwIfDisposed`

If an object is already disposed, the `throwIfDisposed` handle ensures that methods decorated with them throw when disposed.

OrKoN commented on packages/puppeteer-core/src/common/ElementHandle.ts:
```diff
@@ -65,6 +66,7 @@ export class CDPElementHandle<
   }
 
+  @throwIfDisposed()
   override remoteObject(): Protocol.Runtime.RemoteObject {
```
I think this might not need to throw. It does not make any calls and you might still want to have the remote object for the disposed handle.

jrandolf replied:
But is the remote object still available after it has been disposed?

OrKoN replied:
remote object is the metadata about the object, not the object itself.

OrKoN replied:
by looking at the code I don't think the metadata is deleted and I am not sure if it necessarily should be.

jrandolf replied:
The metadata isn't deleted, but it contains information about the handle such as the objectId which is no longer valid after the handle is disposed.

OrKoN replied:
<img width="1397" alt="Screenshot 2023-08-31 at 13 29 40" src="https://github.com/puppeteer/puppeteer/assets/399150/e1c7bbe5-84cc-4482-88ca-5ac696cf486b">

the rest is valid though. E.g., the serialised value of the disposed object.

jrandolf committed:
chore: implement `throwIfDisposed`

If an object is already disposed, the `throwIfDisposed` handle ensures that methods decorated with them throw when disposed.

Lightning00Blade commented on packages/puppeteer-core/src/common/types.ts:
```diff
@@ -29,6 +29,13 @@ export interface Moveable {
+/**
+ * @internal
+ */
+export interface Disposed {
```
Can we/should we create a `super` interface that extend `Disposable`,`AsyncDisposable`,`Disposed`, that way we can be more sure of that the decorator would work in most cases?

jrandolf replied:
Since decorators wrap things, using decorators as a means of composition is considered better than inheritance (see https://www.digitalocean.com/community/tutorials/composition-vs-inheritance).

jrandolf replied:
Discussed offline.
